### PR TITLE
Fix switch accessory & add support for Smart IR (wnykq)

### PR DESCRIPTION
**Pull request by bimusiek**
- Fix bug introduced in https://github.com/0x5e/homebridge-tuya-platform/pull/114
- Add support for temperature & humidity sensor (Smart IR)

### CODE REVIEW

**Comment by bimusiek** [comment]
The error in JSON.parse comes from this response:
```
[11/22/2022, 10:00:50 PM] [TuyaPlatform] [TuyaOpenAPI] Request:
method = get
endpoint = https://openapi.tuyaeu.com
path = /v1.0/devices/bfc4ee178b05afd31dpaai/specifications
query = undefined
headers = {
  "t": "1669150850482",
  "client_id": "k3akuj97e4kxyckwg3yg",
  "nonce": "95f647f0-5a3e-4d8e-93f2-3b8f8f72b364",
  "Signature-Headers": "client_id",
  "sign": "BF8275C52E787797875F5D10B1CA9427512B549DDF4C2A71612435A2C3DB4368",
  "sign_method": "HMAC-SHA256",
  "access_token": "",
  "lang": "en",
  "dev_lang": "javascript",
  "dev_channel": "homebridge",
  "devVersion": "1.6.0-beta.38"
}
body = null
[11/22/2022, 10:00:50 PM] [TuyaPlatform] [TuyaOpenAPI] Response:
path = /v1.0/devices/bfc4ee178b05afd31dpaai/specifications
data = {
  "result": {
    "category": "infrared_light",
    "functions": [
      {
        "code": "BlueLight",
        "type": "STRING",
        "values": "BlueLight"
      },
      {
        "code": "Brightness+",
        "type": "STRING",
        "values": "Brightness+"
      },
      {
        "code": "Brightness-",
        "type": "STRING",
        "values": "Brightness-"
      },
      {
        "code": "ColdLight",
        "type": "STRING",
        "values": "ColdLight"
      },
      {
        "code": "GreenLight",
        "type": "STRING",
        "values": "GreenLight"
      },
      {
        "code": "PowerOff",
        "type": "STRING",
        "values": "PowerOff"
      },
      {
        "code": "PowerOn",
        "type": "STRING",
        "values": "PowerOn"
      },
      {
        "code": "RedLight",
        "type": "STRING",
        "values": "RedLight"
      },
      {
        "code": "WarmLight",
        "type": "STRING",
        "values": "WarmLight"
      }
    ],
    "status": []
  },
  "success": true,
  "t": 1669150850764,
  "tid": "bed90a666aa811ed854422b2ae449af7"
}
```

This opens up the way to support Infrared devices I guess? I will try to make it work soon as I have one light that I would love to be able to control.

How do you feel about adding config option to configure infra devices as this seems like a really custom-tailored list of commands. Or maybe I dont know the inner workings of Tuya platform.
Should I implement this infra light as standard Light and "fake" percentage of brightness? Or make it work as "buttons" so simple switches for each command?

**Comment by 0x5e** [comment]
Thanks for the PR, unfortunately the IR device returns wrong specification format, completely different with others, I have contact the cloud development peoples, due to the history reason they refuse to fix. I prefer to let the plugin convert to the same format before we use. I will confirm how to convert these days.

I prefer add config to override per-device or per-product info, instead of just create a device in the config, so we can keep the device list data source is only from one place (from API). The user can add the IR subdevice in the tuya app, then the IR subdevice can both be used in app and plugin. 

If it should be implemented as the switch accessory or specific accessory depends on the specific IR data the command sent. In your case the brightness command is "Brightness+" "Brightness-" much like an "action", not "Brightness 60%" "Brightness 70%" as a "state".
The "action" should use Switch Accessory, because if use the Light Accessory with brightness slider, when slider goes to the top but the real device is not bright enough, then there's no way to bright the real device :)
The air-condition mostly send the IR "state" data like "27°C" then the HeaterCooler Accessory may be more suitable for them.

Can we open this PR for a while and let me check the spec convertion things? Thanks for your contribution!

File: src/device/TuyaDeviceManager.ts [hunk 1]
```diff
@@ -91,10 +93,11 @@ export default class TuyaDeviceManager extends EventEmitter {
 
     // Combine functions and status together, as it used to be.
     const schemas = new Map<string, TuyaDeviceSchema>();
-    for (const { code, type, values } of [...res.result.status, ...res.result.functions]) {
+    for (const { code, type: rawType, values } of [...res.result.status, ...res.result.functions]) {
       if (schemas[code]) {
         continue;
       }
+      const type = rawType.charAt(0).toUpperCase() + rawType.slice(1).toLowerCase();
```
Comment (0x5e) [inline src/device/TuyaDeviceManager.ts]: My IR device spec is like this:
```
path = /v1.0/devices/6c66dfa384bdc2bbb2yqql/specifications
data = {
  "result": {
    "category": "infrared_ac",
    "functions": [
      {
        "code": "F",
        "type": "ENUM",
        "values": "{\"min\":0,\"max\":3,\"scale\":0,\"step\":1,\"type\":\"Integer\"}"
      },
      {
        "code": "M",
        "type": "ENUM",
        "values": "{\"min\":0,\"max\":4,\"scale\":0,\"step\":1,\"type\":\"Integer\"}"
      },
      {
        "code": "PowerOff",
        "type": "STRING",
        "values": "PowerOff"
      },
      {
        "code": "PowerOn",
        "type": "STRING",
        "values": "PowerOn"
      },
      {
        "code": "T",
        "type": "ENUM",
        "values": "{\"min\":16,\"max\":30,\"scale\":0,\"step\":1,\"type\":\"Integer\"}"
      }
    ],
    "status": [
      {
        "code": "wind",
        "type": "ENUM",
        "values": "{\"min\":0,\"max\":3,\"scale\":0,\"step\":1,\"type\":\"Integer\"}"
      },
      {
        "code": "mode",
        "type": "ENUM",
        "values": "{\"min\":0,\"max\":4,\"scale\":0,\"step\":1,\"type\":\"Integer\"}"
      },
      {
        "code": "power",
        "type": "BOOLEAN",
        "values": "{}"
      },
      {
        "code": "temp",
        "type": "ENUM",
        "values": "{\"min\":16,\"max\":30,\"scale\":0,\"step\":1,\"type\":\"Integer\"}"
      }
    ]
  },
  "success": true,
  "t": 1667552909972,
  "tid": "3fde5b005c2011ed9f2b629ef151a136"
}
```

The `ENUM` actually is `Integer`, the `STRING` is like a `Enum` type with 1 range, `BOOLEAN` is same as `Boolean`. They have to be some dirty code to convert this :(

File: src/platform.ts [hunk 2]
```diff
@@ -256,10 +256,10 @@ export class TuyaPlatform implements DynamicPlatformPlugin {
       this.log.info(`Got home_id=${home_id}, name=${name}`);
       if (this.options.homeWhitelist) {
         if (this.options.homeWhitelist.includes(home_id)) {
-            this.log.info(`Found home_id=${home_id} in whitelist; including devices from this home.`);
-            homeIDList.push(home_id);
+          this.log.info(`Found home_id=${home_id} in whitelist; including devices from this home.`);
```
Comment (0x5e) [inline src/platform.ts:259]: I've noticed this and I don't know why `npm run lint` not return warnings for me and github actions, but `./node_modules/.bin/eslint src/**/**.ts --max-warnings=0` works. Do you know why?

**Comment by bimusiek** [comment]
So just to clarify things, Smart IR reports temp and humidity with proper schema. It behaves like thermometer so I think its fine to make it be discovered as one. 

Other ir devices could be parsed as each action is a switch but I could prepare separate PR for that.

File: src/accessory/AccessoryFactory.ts [hunk 3]
```diff
@@ -104,6 +104,7 @@ export default class AccessoryFactory {
       case 'co2bj':
         handler = new CarbonDioxideSensorAccessory(platform, accessory);
         break;
+      case 'wnykq':
       case 'wsdcg':
         handler = new TemperatureHumiditySensorAccessory(platform, accessory);
```
Comment (0x5e) [inline src/accessory/AccessoryFactory.ts:112]: `TemperatureHumiditySensorAccessory` claims to require one of the following schemas: `['va_temperature', 'va_humidity', 'humidity_value']`, and not all Smart IR have temperature or humidity sensors (mines don't have these sensor), the plugin will raise the warnings for those device who doesn't have the required schema. This is intend for find out the device which is "non-standard" (lossing some schema).

So maybe we can subclass the TemperatureHumiditySensorAccessory and override `requiredSchema()` to return an empty array.

Currently there's not too much convient to reuse the code across different accessories, only subclass and copy the code, I will improve later.

File: src/device/TuyaDeviceManager.ts [hunk 4]
```diff
@@ -107,7 +110,11 @@ export default class TuyaDeviceManager extends EventEmitter {
       }
       let property: TuyaDeviceSchemaProperty;
       try {
-        property = JSON.parse(values);
+        if (type === TuyaDeviceSchemaType.String) {
```
Comment (0x5e) [inline src/device/TuyaDeviceManager.ts]: `STRING` value is a string, but `String` value is a json object string like `{\"maxlen\":255}` defines the max length.

**Comment by 0x5e** [comment]
Sorry I forget to submit the review, It was left as pending state and you can't see. Now it is visible.

**Comment by 0x5e** [comment]
Here's some informations:

IR additional API, if IoT Core not satisfied your requirements:
https://developer.tuya.com/en/docs/cloud/ir-control-hub-open-service?id=Kb3oe2mk8ya72

A Tuya plugin focused on IR devices:
https://github.com/prasad-edlabadka/homebridge-tuya-ir

Thanks again, I was intend to do this after December, but I'm a little bit busy and don't have enough time to develop and test recently.

File: src/platform.ts (hunk 2 above)
Comment (bimusiek) [reply src/platform.ts:259]: In my case, using eslint directly also works:
```
❯ ./node_modules/.bin/eslint src/**/**.ts --max-warnings=0
❯ echo $?
0
```
Do you have some additional global config specified? Or some prettier rules if you are using VSCode?

Comment (bimusiek) [reply src/platform.ts:259]: Shouldnt it be 
```
eslint src/**/*.ts --max-warnings=0
```
? with one asterisk before `.ts`

Besides this, no clue why it wouldnt return errors in actions.

File: src/device/TuyaDeviceManager.ts (hunk 1 above)
Comment (bimusiek) [reply src/device/TuyaDeviceManager.ts]: Thanks for explaining, I will revert this code for now and we can discuss it further what is the best approach here

**Comment by bimusiek** [comment]
Hey @0x5e please re-check 👍

**Review by 0x5e — APPROVED** [review]

**Comment by 0x5e** [comment]
Thanks again! @bimusiek

**Pull request merged**